File: apps/ios/ExpCore/Sources/Auth/AccountStore.swift
import Foundation

private let keyAccounts = "accounts"
private let keyActiveAccountId = "active_account_id"
// One-shot flag: URL-keyed accounts have been re-keyed to per-user ids.
private let keyPerUserMigrationDone = "peruser_migration_v1"

// Legacy single-account keys, migrated on first launch.
private let legacyKeyInstanceUrl = "instance_url"
private let legacyKeyToken = "session_token"
private let legacyKeyUserEmail = "user_email"
private let legacyKeyUserName = "user_name"
private let legacyKeyUserId = "user_id"
private let legacyKeyIsAdmin = "is_admin"

public final class AccountStore: @unchecked Sendable {
    private let store: any KeychainStoring
    private let lock = NSLock()
    private var cached: [ServerAccount]
    private var cachedActiveId: String?

    public init(keychain: any KeychainStoring) {
        self.store = keychain
        self.cached = AccountStore.loadAccounts(from: keychain)
        self.cachedActiveId = keychain.get(keyActiveAccountId)
        AccountStore.migrateLegacyIfNeeded(store: keychain, accounts: &self.cached, activeId: &self.cachedActiveId)
        AccountStore.migratePerUserIdsIfNeeded(store: keychain, accounts: &self.cached, activeId: &self.cachedActiveId)
        persist()
    }

    public var accounts: [ServerAccount] {
        lock.withLock { cached }
    }

    public var activeAccount: ServerAccount? {
        lock.withLock {
            guard let id = cachedActiveId else { return nil }
            return cached.first { $0.id == id }
        }
    }

    public var activeAccountId: String? {
        lock.withLock { cachedActiveId }
    }

    /// Adds (or merges) an account by instance URL and marks it active.
    /// Returns the resulting account, with a stable id.
    @discardableResult
    public func upsertAndActivate(instanceUrl: String) -> ServerAccount {
        lock.lock()
        defer { lock.unlock() }
        let id = ServerAccount.makeId(for: instanceUrl)
        if let idx = cached.firstIndex(where: { $0.id == id }) {
            cached[idx].instanceUrl = instanceUrl
            cached[idx].lastUsedAt = Date()
        } else {
            cached.append(ServerAccount(
                id: id,
                instanceUrl: instanceUrl,
                token: nil,
                userEmail: nil,
                userName: nil,
                userId: nil,
                isAdmin: false,
                lastUsedAt: Date()
            ))
        }
        cachedActiveId = id
        persistLocked()
        return cached.first { $0.id == id }!
    }

    /// Resolve which account a successful login belongs to (per-user identity)
    /// and persist its token + fields. Implements the account state machine:
    /// - same user on the active server → in-place token refresh (no DB wipe);
    /// - a per-user account that already exists → switch to it (offline cache
    ///   preserved), dropping the tokenless pending record we signed in from;
    /// - a pending (pre-login) record → re-keyed to the per-user id, so the
    ///   fresh DB file triggers a full snapshot sync;
    /// - another user already resolved on this server → a fresh per-user record
    ///   rather than clobbering theirs.
    /// The caller MUST pass a resolved userId (the login VMs fail the login if a
    /// session never resolves one). Returns the resolved account id.
    @discardableResult
    public func resolveActiveAccount(
        token: String,
        email: String?,
        name: String?,
        userId: String,
        isAdmin: Bool,
        onboardingCompletedAt: String?,
        onboardingKnown: Bool
    ) -> String {
        lock.lock()
        defer { lock.unlock() }

        guard let activeId = cachedActiveId,
              let activeIdx = cached.firstIndex(where: { $0.id == activeId }) else {
            return cachedActiveId ?? ""
        }
        let instanceUrl = cached[activeIdx].instanceUrl
        let perUserId = ServerAccount.makeId(instanceUrl: instanceUrl, userId: userId)

        if let existingIdx = cached.firstIndex(where: { $0.id == perUserId }) {
            applyResolvedFields(
                at: existingIdx, token: token, email: email, name: name,
                userId: userId, isAdmin: isAdmin,
                onboardingCompletedAt: onboardingCompletedAt, onboardingKnown: onboardingKnown
            )
            cachedActiveId = perUserId
            // Drop the distinct tokenless/userless pending record we came from.
            if activeId != perUserId, cached[activeIdx].token == nil, cached[activeIdx].userId == nil {
                cached.removeAll { $0.id == activeId }
            }
            persistLocked()
            return perUserId
        }

        if cached[activeIdx].userId == nil {
            // Pending (pre-login) record → re-key to the per-user id.
            cached[activeIdx].id = perUserId
            applyResolvedFields(
                at: activeIdx, token: token, email: email, name: name,
                userId: userId, isAdmin: isAdmin,
                onboardingCompletedAt: onboardingCompletedAt, onboardingKnown: onboardingKnown
            )
            cachedActiveId = perUserId
        } else {
            // Another user is already resolved on this server — create a fresh
            // per-user record instead of clobbering theirs.
            cached.append(ServerAccount(
                id: perUserId,
                instanceUrl: instanceUrl,
                token: token,
                userEmail: email,
                userName: name,
                userId: userId,
                isAdmin: isAdmin,
                onboardingCompletedAt: onboardingCompletedAt,
                onboardingKnown: onboardingKnown ? true : nil,
                lastUsedAt: Date()
            ))
            cachedActiveId = perUserId
        }
        persistLocked()
        return perUserId
    }

    /// Assign the resolved login fields onto `cached[idx]`, preserving a prior
    /// onboarding flag when the incoming session didn't report one (a transient
    /// session read must never bounce a returning user back to the wizard).
    /// Assumes the store lock is already held.
    private func applyResolvedFields(
        at idx: Int, token: String, email: String?, name: String?, userId: String,
        isAdmin: Bool, onboardingCompletedAt: String?, onboardingKnown: Bool
    ) {
        let priorCompleted = cached[idx].onboardingCompletedAt
        let priorKnown = cached[idx].onboardingKnown
        cached[idx].token = token
        cached[idx].userEmail = email
        cached[idx].userName = name
        cached[idx].userId = userId
        cached[idx].isAdmin = isAdmin
        cached[idx].onboardingCompletedAt = onboardingCompletedAt ?? priorCompleted
        cached[idx].onboardingKnown = (onboardingKnown || priorKnown == true) ? true : nil
        cached[idx].lastUsedAt = Date()
    }

    /// Marks an account onboarded (after onboarding.complete succeeds) so the
    /// nav gate stops showing the wizard without waiting for a session re-read.
    public func setOnboardingCompleted(id: String, completedAtIso: String) {
        lock.lock()
        defer { lock.unlock() }
        guard let idx = cached.firstIndex(where: { $0.id == id }) else { return }
        cached[idx].onboardingCompletedAt = completedAtIso
        cached[idx].onboardingKnown = true
        persistLocked()
    }

    /// Switches the active account to the given id. No-op if id is unknown.
    public func setActive(id: String) {
        lock.lock()
        defer { lock.unlock() }
        guard cached.contains(where: { $0.id == id }) else { return }
        cachedActiveId = id
        if let idx = cached.firstIndex(where: { $0.id == id }) {
            cached[idx].lastUsedAt = Date()
        }
        persistLocked()
    }

    /// Removes an account. If the removed account was active, picks the most-recent remaining
    /// account as the new active, or nil if none remain.
    public func remove(id: String) {
        lock.lock()
        defer { lock.unlock() }
        cached.removeAll { $0.id == id }
        if cachedActiveId == id {
            cachedActiveId = cached.max(by: { $0.lastUsedAt < $1.lastUsedAt })?.id
        }
        persistLocked()
    }

    // MARK: - Persistence

    private func persist() {
        lock.lock()
        defer { lock.unlock() }
        persistLocked()
    }

    private func persistLocked() {
        if let data = try? JSONEncoder().encode(cached), let json = String(data: data, encoding: .utf8) {
            store.set(keyAccounts, value: json)
        }
        store.set(keyActiveAccountId, value: cachedActiveId)
    }

    private static func loadAccounts(from keychain: any KeychainStoring) -> [ServerAccount] {
        guard
            let json = keychain.get(keyAccounts),
            let data = json.data(using: .utf8),
            let accounts = try? JSONDecoder().decode([ServerAccount].self, from: data)
        else { return [] }
        return accounts
    }

    private static func migrateLegacyIfNeeded(
        store: any KeychainStoring,
        accounts: inout [ServerAccount],
        activeId: inout String?
    ) {
        // If we already have any accounts, skip migration.
        guard accounts.isEmpty else { return }
        guard let legacyUrl = store.get(legacyKeyInstanceUrl) else { return }
        let id = ServerAccount.makeId(for: legacyUrl)
        let account = ServerAccount(
            id: id,
            instanceUrl: legacyUrl,
            token: store.get(legacyKeyToken),
            userEmail: store.get(legacyKeyUserEmail),
            userName: store.get(legacyKeyUserName),
            userId: store.get(legacyKeyUserId),
            isAdmin: store.get(legacyKeyIsAdmin) == "true",
            lastUsedAt: Date()
        )
        accounts = [account]
        activeId = id
        store.delete(legacyKeyInstanceUrl)
        store.delete(legacyKeyToken)
        store.delete(legacyKeyUserEmail)
        store.delete(legacyKeyUserName)
        store.delete(legacyKeyUserId)
        store.delete(legacyKeyIsAdmin)
    }

    // One-shot re-key of URL-keyed accounts to per-user ids. Before this, two
    // users on the same server shared one accountId (hence one DB file → "logged
    // into the wrong account"). Re-key each signed-in URL-keyed record to
    // makeId(instanceUrl:userId:); a signed-in record with no captured userId
    // gets its token nulled to force a clean re-login. The stale URL-keyed DB
    // files are wiped app-side (AppDependencies) — the cache may be the wrong
    // user's data, so it's deleted, not renamed. Guarded by a persisted flag;
    // idempotent (the keychain survives reinstall).
    static func migratePerUserIdsIfNeeded(
        store: any KeychainStoring,
        accounts: inout [ServerAccount],
        activeId: inout String?
    ) {
        guard store.get(keyPerUserMigrationDone) != "true" else { return }
        for idx in accounts.indices {
            let account = accounts[idx]
            // Only touch legacy URL-keyed records; per-user ids already differ.
            guard account.id == ServerAccount.makeId(for: account.instanceUrl) else { continue }
            guard account.token != nil else { continue }
            if let userId = account.userId, !userId.isEmpty {
                let perUserId = ServerAccount.makeId(instanceUrl: account.instanceUrl, userId: userId)
                if perUserId != account.id {
                    if activeId == account.id { activeId = perUserId }
                    accounts[idx].id = perUserId
                }
            } else {
                // Signed-in but no userId — can't derive a per-user id. Null the
                // token so the next launch re-authenticates cleanly.
                accounts[idx].token = nil
            }
        }
        store.set(keyPerUserMigrationDone, value: "true")
    }
}

File: apps/android/app/src/main/java/com/exponential/app/data/auth/AccountStore.kt
package com.exponential.app.data.auth

import javax.inject.Inject
import javax.inject.Singleton
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.serialization.json.Json

private const val KEY_ACCOUNTS = "accounts"
private const val KEY_ACTIVE_ACCOUNT_ID = "active_account_id"
// One-shot re-key of URL-only account ids to per-user ids (see migratePerUserIdsLocked).
private const val KEY_PERUSER_MIGRATION = "peruser_migration_v1"

// Legacy single-account keys migrated on first launch.
private const val LEGACY_KEY_INSTANCE_URL = "instance_url"
private const val LEGACY_KEY_TOKEN = "session_token"
private const val LEGACY_KEY_USER_EMAIL = "user_email"
private const val LEGACY_KEY_USER_ID = "user_id"
private const val LEGACY_KEY_IS_ADMIN = "is_admin"

@Singleton
class AccountStore @Inject constructor(
    private val store: SecureStore,
) {
    private val json = Json { ignoreUnknownKeys = true; encodeDefaults = true }
    private val lock = Any()

    private val _accounts = MutableStateFlow<List<ServerAccount>>(emptyList())
    val accounts: StateFlow<List<ServerAccount>> = _accounts.asStateFlow()

    private val _activeAccountId = MutableStateFlow<String?>(null)
    val activeAccountId: StateFlow<String?> = _activeAccountId.asStateFlow()

    init {
        synchronized(lock) {
            val loaded = loadAccounts()
            val activeId = store.get(KEY_ACTIVE_ACCOUNT_ID)
            val (migratedAccounts, migratedActive) = migrateLegacyIfNeeded(loaded, activeId)
            _accounts.value = migratedAccounts
            _activeAccountId.value = migratedActive
            migratePerUserIdsLocked()
            persistLocked()
        }
    }

    val activeAccount: ServerAccount?
        get() = _activeAccountId.value?.let { id -> _accounts.value.firstOrNull { it.id == id } }

    fun upsertAndActivate(instanceUrl: String): ServerAccount {
        synchronized(lock) {
            val id = ServerAccount.makeId(instanceUrl)
            val now = System.currentTimeMillis()
            val existing = _accounts.value.firstOrNull { it.id == id }
            val updated = if (existing != null) {
                _accounts.value.map { if (it.id == id) it.copy(instanceUrl = instanceUrl, lastUsedAt = now) else it }
            } else {
                _accounts.value + ServerAccount(id = id, instanceUrl = instanceUrl, lastUsedAt = now)
            }
            _accounts.value = updated
            _activeAccountId.value = id
            persistLocked()
            return updated.first { it.id == id }
        }
    }

    // Persist a resolved login. `userId` is REQUIRED (the login flow fails
    // rather than call this with an unknown user): it drives the per-user
    // account identity via [resolveAccounts], which re-keys the active pending
    // account to the per-user id (fresh DB) or merges into the existing per-user
    // account (its cached DB resumes) — never sharing a DB across users.
    fun resolveActiveAccount(
        token: String,
        email: String?,
        name: String?,
        userId: String,
        isAdmin: Boolean,
        onboardingCompletedAt: String?,
        onboardingKnown: Boolean,
    ) {
        synchronized(lock) {
            val active = _activeAccountId.value?.let { id -> _accounts.value.firstOrNull { it.id == id } }
            val instanceUrl = active?.instanceUrl ?: return
            val result = resolveAccounts(
                accounts = _accounts.value,
                instanceUrl = instanceUrl,
                userId = userId,
                token = token,
                email = email,
                name = name,
                isAdmin = isAdmin,
                onboardingCompletedAt = onboardingCompletedAt,
                onboardingKnown = onboardingKnown,
                now = System.currentTimeMillis(),
            )
            _accounts.value = result.accounts
            _activeAccountId.value = result.activeId
            persistLocked()
        }
    }

    // Flip just the onboarding flag on a given account (after onboarding.complete),
    // leaving the token/session fields intact.
    fun setOnboardingCompletedAt(id: String, value: String?) {
        synchronized(lock) {
            _accounts.value = _accounts.value.map {
                if (it.id == id) it.copy(onboardingCompletedAt = value, onboardingKnown = true) else it
            }
            persistLocked()
        }
    }

    fun clearActiveToken() {
        synchronized(lock) {
            val id = _activeAccountId.value ?: return
            _accounts.value = _accounts.value.map {
                if (it.id == id) it.copy(token = null) else it
            }
            persistLocked()
        }
    }

    fun setActive(id: String) {
        synchronized(lock) {
            if (_accounts.value.none { it.id == id }) return
            val now = System.currentTimeMillis()
            _accounts.value = _accounts.value.map {
                if (it.id == id) it.copy(lastUsedAt = now) else it
            }
            _activeAccountId.value = id
            persistLocked()
        }
    }

    fun remove(id: String) {
        synchronized(lock) {
            _accounts.value = _accounts.value.filterNot { it.id == id }
            if (_activeAccountId.value == id) {
                _activeAccountId.value = _accounts.value.maxByOrNull { it.lastUsedAt }?.id
            }
            persistLocked()
        }
    }

    private fun persistLocked() {
        store.set(KEY_ACCOUNTS, json.encodeToString(_accounts.value))
        store.set(KEY_ACTIVE_ACCOUNT_ID, _activeAccountId.value)
    }

    private fun loadAccounts(): List<ServerAccount> {
        val raw = store.get(KEY_ACCOUNTS) ?: return emptyList()
        return runCatching {
            json.decodeFromString<List<ServerAccount>>(raw)
        }.getOrElse { emptyList() }
    }

    // One-shot: re-key accounts stored under the old URL-only id to the per-user
    // id, so each user gets a distinct Room DB file. A token without a userId
    // (older build that never captured one) is cleared to force a clean re-login
    // rather than guess an identity. The legacy URL-keyed DB files are wiped
    // app-side (see the per-user DB cleanup in ExponentialApp).
    private fun migratePerUserIdsLocked() {
        if (store.get(KEY_PERUSER_MIGRATION) == "done") return
        var activeId = _activeAccountId.value
        _accounts.value = _accounts.value.map { account ->
            val uid = account.userId
            when {
                account.token != null && uid != null &&
                    account.id == ServerAccount.makeId(account.instanceUrl) -> {
                    val newId = ServerAccount.makeId(account.instanceUrl, uid)
                    if (activeId == account.id) activeId = newId
                    account.copy(id = newId)
                }
                account.token != null && uid == null -> account.copy(token = null)
                else -> account
            }
        }
        _activeAccountId.value = activeId
        store.set(KEY_PERUSER_MIGRATION, "done")
    }

    private fun migrateLegacyIfNeeded(
        loaded: List<ServerAccount>,
        activeId: String?,
    ): Pair<List<ServerAccount>, String?> {
        if (loaded.isNotEmpty()) return loaded to activeId
        val legacyUrl = store.get(LEGACY_KEY_INSTANCE_URL) ?: return emptyList<ServerAccount>() to null
        val id = ServerAccount.makeId(legacyUrl)
        val account = ServerAccount(
            id = id,
            instanceUrl = legacyUrl,
            token = store.get(LEGACY_KEY_TOKEN),
            userEmail = store.get(LEGACY_KEY_USER_EMAIL),
            userId = store.get(LEGACY_KEY_USER_ID),
            isAdmin = store.get(LEGACY_KEY_IS_ADMIN) == "true",
        )
        // Wipe legacy keys.
        store.set(LEGACY_KEY_INSTANCE_URL, null)
        store.set(LEGACY_KEY_TOKEN, null)
        store.set(LEGACY_KEY_USER_EMAIL, null)
        store.set(LEGACY_KEY_USER_ID, null)
        store.set(LEGACY_KEY_IS_ADMIN, null)
        return listOf(account) to id
    }
}
